src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java

```diff
@@ -0,0 +1,444 @@
+            .andShould()
+            .beFinal()
+            .andShould()
+            .beStatic()
+            .orShould(beSuppressedField);
+
+        fieldsShouldBeImmutable.check(utilClasses);
+    }
+
+    /**
+     * Test to ensure modules annotated with {@link StatelessCheck} contain immutable fields
+     * and modules with mutable fields are annotated with {@link FileStatefulCheck} or
+     * {@link GlobalStatefulCheck}.
+     *
+     * @noinspection JUnitTestMethodWithNoAssertions
+     * @noinspectionreason JUnitTestMethodWithNoAssertions - asserts in callstack,
+     *     but not in this method
+     */
+    @Test
+    public void testModuleImmutability() {
+        final JavaClasses checkstyleChecks = new ClassFileImporter()
+            .withImportOption(ImportOption.Predefined.DO_NOT_INCLUDE_TESTS)
+            .importPackages("com.puppycrawl.tools.checkstyle.checks");
+
+        final DescribedPredicate<JavaField> moduleProperties = new ModulePropertyPredicate();
+
+        final DescribedPredicate<JavaClass> immutableFields = new ImmutableFieldsPredicate();
+
+        final ArchCondition<JavaClass> beSuppressedClass = new SuppressionArchCondition(
+                SUPPRESSED_CLASSES, "be suppressed");
+
+        final ArchCondition<JavaField> beSuppressedField = new SuppressionArchCondition<>(
+            SUPPRESSED_FIELDS_IN_MODULES, "be suppressed");
+
+        final ArchCondition<JavaField> bePrivateExceptMessageKeys =
+            new ArchCondition<>("be private") {
+
+                @Override
+                public void check(JavaField item, ConditionEvents events) {
+                    if (!item.getName().startsWith("MSG")
+                        && !item.getModifiers().contains(JavaModifier.PRIVATE)) {
+                        final String message = String
+                            .format(Locale.ROOT, "Field <%s> should %s in %s",
+                                    item.getFullName(), getDescription(),
+                                    item.getSourceCodeLocation().toString());
+                        events.add(SimpleConditionEvent.violated(item, message));
+                    }
+                }
+            };
+
+        final ArchRule fieldsInStatelessChecksShouldBeImmutable = fields()
+            .that()
+            .areDeclaredInClassesThat()
+            .areAnnotatedWith(StatelessCheck.class)
+            .and(are(not(moduleProperties)))
+            .should(BE_IMMUTABLE)
+            .andShould()
+            .beFinal()
+            .andShould(bePrivateExceptMessageKeys)
+            .orShould(beSuppressedField);
+
+        final ArchRule classesWithImmutableFieldsShouldBeStateless = classes()
+            .that(have(immutableFields))
+            .and()
+            .haveSimpleNameEndingWith("Check")
+            .and()
+            .doNotHaveModifier(JavaModifier.ABSTRACT)
+            .should()
+            .beAnnotatedWith(StatelessCheck.class)
+            .orShould(beSuppressedClass);
```

Comment:
Let's have two separate suppression lists, one for this rule and one for the below rule. We can use both lists in a given rule; we can filter by the 'other' list first with some ArchCondition like `shouldBeAnnotatedWith*OtherAnnotation*`.

Reply:
Created separate lists.